src/app/shell/item-comparators.ts

```diff
@@ -109,6 +110,114 @@ const TAG_ITEM_COMPARATORS: {
   archive: (getTag) => compareBy((item) => getTag(item) === 'archive'),
 };
 
+export type VaultGroupValue = string | number | boolean | undefined;
+
+interface MutableVaultGroup {
+  groupingValue: VaultGroupValue;
+  iconValue?: VaultGroupIconValue;
+  items: DimItem[];
+}
+
+interface VaultGroup {
+  groupingValue: VaultGroupValue;
+  iconValue?: VaultGroupIconValue;
+  items: readonly DimItem[];
+}
+
+const GROUPING_VALUE_GETTERS: {
+  [key: string]: (
+    item: DimItem,
+    getTag: (item: DimItem) => TagValue | undefined,
+  ) => VaultGroupValue;
+} = {
+  tag: (item, getTag) => getTag(item),
+  typeName: (item) => item.typeName,
+  rarity: (item) => D2ItemTiers[item.tier],
+  ammoType: (item) => item.ammoType,
+  elementWeapon: (item) => {
+    if (item.bucket.inWeapons) {
+      return item.element?.enumValue ?? Number.MAX_SAFE_INTEGER;
+    }
+  },
+};
+
+export type VaultGroupIconValue =
+  | TagValue
+  | DestinyDamageTypeDefinition
+  | DestinyAmmunitionType
+  | number[]
+  | undefined
+  | null;
```

Comment:
It would be good to avoid `null` here in favor of `undefined` - handling both makes things unnecessarily complicated.

Reply:
I opted for `null` in addition to `undefined` so that items that explicitly don't have an icon have `null`, whereas items that should have an icon but it is missing result in `undefined`. Just makes debugging a little easier.

Reply:
I generally try hard to avoid encoding semantic meaning in the difference between `null`/`undefined`

Reply:
I guess this will also be removed when I move to tagged union.